Add "Upload SIPs" page to Dashboard (#1186)

djjuhasz opened the pull request:
- Return a "202 Created" HTTP success code from API "/ingest/sips/upload" endpoint
- Add an "Upload SIPs" page to the Dashboard using the Uppy file upload library

djjuhasz committed:
Add Upload SIPs page to Dashboard (#1186)

Add an "Upload SIPs" page to the Dashboard using the Uppy file upload
library.

jraddaoui commented on dashboard/src/pages/ingest/upload/index.vue:
```diff
@@ -0,0 +1,54 @@
+
+import "@uppy/core/dist/style.min.css";
+import "@uppy/dashboard/dist/style.css";
+import "@uppy/progress-bar/dist/style.min.css";
```
I wouldn't use the minified version, it could help on development and the build process should minify them.

djjuhasz replied:
Ah, okay.

jraddaoui commented on dashboard/src/pages/ingest/upload/index.vue:
```diff
@@ -0,0 +1,54 @@
+  allowedMetaFields: false,
+  getResponseData: () => {
+    return { url: "" };
+  },
```
Is this really needed?

djjuhasz replied:
Yes, Uppy expects a JSON response containing a "URL" field or it shows the upload as "failed". My guess is Uppy wants to provide a link to the uploaded resource.  I think we should replace this with an actual URL after #1210 is implemented.

Uppy has a lot of options, so it's possible that there's a way to turn off the expectation of a URL in the success response, but I haven't found one.

jraddaoui commented on dashboard/src/pages/ingest/upload/index.vue:
```diff
@@ -0,0 +1,54 @@
+        });
+      }, 500);
+    }
+  },
```
This will redirect after the first response is received, but why the timeout?

djjuhasz replied:
I found the immediate redirect after the upload completes to be jarring, and I prefer the brief pause.  We can see how this tests, and adjust or eliminate the timeout if the feedback is negative.

I think once #1210 is implemented we can return a URL for the uploaded SIP and remove this redirect altogether.

jraddaoui commented on dashboard/src/pages/ingest/upload/index.vue:
```diff
@@ -0,0 +1,54 @@
+      }, 500);
+    }
+  },
+});
```
Do we want to do something like what they suggest to refresh the access token if needed?

https://uppy.io/docs/xhr-upload/#how-can-i-refresh-auth-tokens-after-they-expire

We sign out the user directly in other requests:

https://github.com/artefactual-sdps/enduro/blob/main/dashboard/src/client.ts#L68-L73

Not sure what would happen here. Would it should an error? Would it continue retrying?

djjuhasz replied:
Yeah, I think this is a good idea -- I've gotten an error a few times trying to upload with an expired token.

djjuhasz replied:
I've implemented the `authStore.signinSilent()` method to refresh the access token if needed. This solution works for my tests instead of failing if the token is expired, as it used to do.  It's hard to manually test all the possible scenarios though, and I'm not sure how to do automated testing of how the uploader handles an expired token. :(

I tried to add a `catch()` to the token renewal process in case `authStore.signinSilent()` fails, but I didn't see any difference in the uploader behaviour, and again I'm not sure how to automate testing this functionality.

jraddaoui replied:
Nice, thanks @djjuhasz! Looks like a good approach, I'd like to revisit how we do this in other API interactions, so we can go from this.

> I tried to add a catch() to the token renewal process in case authStore.signinSilent() fails, but I didn't see any difference in the uploader behaviour, and again I'm not sure how to automate testing this functionality.

I think the issue is that we don't return the promise from `authStore.signinSilent()`, I think we should deal with the error from the manager there, no need to do it now.

jraddaoui commented on dashboard/src/pages/ingest/upload/index.vue:
```diff
@@ -0,0 +1,54 @@
+
+<template>
+  <div class="container-xxl">
+    <h1><IconUpload class="me-3 text-dark" />Upload SIPs</h1>
```
Other pages use `class="d-flex"` in the main heading, I think that may align the icon and text better.

jraddaoui commented on internal/api/design/ingest.go:
```diff
@@ -214,6 +214,8 @@ var _ = Service("ingest", func() {
 			// direct access to the HTTP request body reader.
 			SkipRequestBodyEncodeDecode()
 
+			Response(StatusCreated)
```
I'd use `202 Accepted` instead of `201 Created`, there is a lot to happen until the request is completed and the resources created. Even if we are not able yet to return an identifier or URL to track the status, we should be able to do so after #1209 and #1210 are completed.

djjuhasz replied:
Yeah, fair enough.  I debated 201 vs. 202 but I think you're right that 202 Accepted is more appropriate right now.

jraddaoui reviewed (commented):
Looking great @djjuhasz!

djjuhasz committed:
Return a "202 Accepted" response on upload

Return a "202 Accepted" response code from the "/ingest/sips/upload"
endpoint on successful upload.

jraddaoui commented on dashboard/src/pages/ingest/upload/index.vue:
```diff
@@ -0,0 +1,72 @@
+  allowedMetaFields: false,
+  // Called again for every retry too.
+  async onBeforeRequest(xhr) {
+    if (!authStore.getUserAccessToken) {
```
I'd use `authStore.isUserValid`, which also checks if the token is expired.

jraddaoui commented on dashboard/src/pages/ingest/upload/index.vue:
```diff
@@ -0,0 +1,72 @@
+          });
+        }, 500);
+        break;
+      // "401 Unauthorized" is returned when the auth token has expired.
+      case 401:
+        await authStore.signinSilent();
+        break;
```
If we do it before request, why do we need it after response?

djjuhasz replied:
That's the way they do it in https://uppy.io/docs/xhr-upload/#how-can-i-refresh-auth-tokens-after-they-expire

I guess theoretically the token could expire between the pre-request check and the request actually reaching the server. :shrug:

jraddaoui replied:
But doing it after response won't fix that issue. If `onBeforeRequest` is called on retries too (as they said), it doesn't make sense to me to do it in `onAfterResponse`.

jraddaoui reviewed (approved):
Thanks @djjuhasz!
